### PR TITLE
✨ [#1611] Add setup configuration step for authorization

### LINKED ISSUE #1611: Allow CICD-configuration for authorisations
### Thema / Theme

Other

### Omschrijving / Description

**Probleem**

1. Technische configuratie-items zijn niet via CI/CD pipeline te besturen:
• Dit zorgt ervoor dat alle configuratie handmatig moet worden uitgevoerd
• Elke gemeente moet separaat en handmatig worden geconfigureerd met veel deplicatie
• Het gevolg is dat Dimpact niet schaalbaar kan zijn als beheerorganisatie

2. Technisch en functioneel beheer zijn onvoldoende gescheiden
• Dit is een probleem aangezien gevolgen van aanpassingen voor technisch beheer groot zijn
(risico voor beschikbaarheid applicaties)
• Niet elke beheerder moet technisch beheer rechten krijgen en voor sommige gemeentes zal
alleen Dimpact het technisch beheer doen

**Acceptance criteria**

1. Technische configuratie kan programmatisch worden ingesteld vanuit de CI/CD pipeline
2. Dit is van toepassing op alle componenten die in gebruik zijn binnen PodiumD
3. Technisch beheer kan nog steeds via een GUI worden aangepast
4. Middels rollen en rechten is technisch beheer gescheiden van functioneel beheer

**Oplossingsrichting**

1. Er wordt van elk component een inventarisatie gemaakt welke environment variabelen wel/niet
ondersteund worden.
2. Er wordt overkoepelend bepaald welke environment variabelen nog moeten worden
toegevoegd.
3. Een aantal environment variabelen zal effect hebben op in de database opgeslagen waarden. De
database waarden worden overschreven met de waarden uit de environment variabelen.
4. Er komt in alle componenten een “setup_configuration” commando, waarmee environment
variabelen worden opgeslagen in de database. Dit commando kan middels een Kubernetes job
worden aangeroepen, met een init-container, of handmatig.
5. Een aantal configuratie instellingen zijn dermate uitgebreid dat deze niet met eenvoudige key-
value (standaard environment variabelen) methodiek kunnen worden opgegeven. Daarom zijn
sommige waarden middels een YAML-bestand in te laden, waarnaar verwezen kan worden met
een environment variabele. Het YAML-bestand zelf kan een gemounte ConfigMap zijn.
6. Environment variabelen die database waarden overschrijven die ook in de beheer UI zijn aan te
passen, blijven aanpasbaar in de beheer UI. Middels rechten kan je deze alleen-lezen maken. We
gebruiken hiervoor het standaard rechten-model in het technische framework, en bouwen geen
laag daar boven op.
7. Het einddoel is om Open Zaak en Open Notificaties volledig werkend en geconfigureerd te
hebben op een Kubernetes cluster, in het bijzonder op PodiumD, zonder dat in de
beheerinterface hoeft te worden ingelogd.

### Toegevoegde waarde / Added value

No more admin UI after deploy/update

### Aanvullende opmerkingen / Additional context

_No response_

### LINKED ISSUE #1655: Add documentation for authorization setup configuration step
Feature is implemented in https://github.com/open-zaak/open-zaak/pull/1624
Related issue: https://github.com/open-zaak/open-zaak/issues/1611

describe:
- [x] dumping config
- [x] describe that it's important to use the same uuids when importing/exporting catalogi
- [x] loading config
- [x] mapping for domains
- [x] describe behavior of regular vs --overwrite (also what happens in case of collisions)
  * no overwrite: if all applicatie UUIDs and JWTsecret identifiers are present, no changes are made (todo check what happens if one of them is not present, will it update everything?)
  * overwrite: update existing entries

### PR DESCRIPTION
**Pull request by stevenbal**
Fixes #1611
Fixes #1655 
Requires https://github.com/maykinmedia/commonground-api-common/pull/27 to be merged and released

**Changes**
* Add command to dump authorization configuration to a file
* Add configuration step to take this file as input (together with a domain mapping file) and import the authorization setup
* Add documentation

### CODE REVIEW

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [inline src/openzaak/config/bootstrap/authorizations.py:90]: I'm not sure how we would test if this works, it seems way too excessive to try and access every zaaktype in the autorisaties we just loaded :sweat_smile:

**Comment by annashamray** [comment]
You can add the doc in the separate issue, let's just not forget about it

**Comment by stevenbal** [comment]
created a separate issue for the docs: https://github.com/open-zaak/open-zaak/issues/1655

File: src/openzaak/management/commands/dump_auth_fixtures.py
Comment (annashamray) [inline src/openzaak/management/commands/dump_auth_fixtures.py:1]: You know it looks like you just reuse the existing command without modifying it. I think it could better be a  bash file, just like `bin/generate_admin_index_fixture.sh`, which we already have.

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (annashamray) [inline src/openzaak/config/tests/bootstrap/test_load_authorizations.py:41]: I think it could be nice to have a real `auth_dump.yaml` file added to the test folder, so we could see its content as an example. The same applies for `mapping.yaml` file

File: src/openzaak/config/bootstrap/authorizations.py
Comment (annashamray) [inline src/openzaak/config/bootstrap/authorizations.py:71]: Have you discussed the shape of mapping file with Joeri or Sjoerd? It looks like a bit of overkill, it could be just an env var `AUTHORIZATIONS_CONFIG_TARGET_DOMAIN` or something like this
On the other hand we make these commands for devops people. so I don't know what it more convenient for them

File: src/openzaak/management/commands/dump_auth_fixtures.py
Comment (annashamray) [inline src/openzaak/management/commands/dump_auth_fixtures.py:29]: TIL that you can generate fixtures in yaml out of the box

**Review by annashamray — CHANGES_REQUESTED** [review]
It looks very neat and nice! 

1. For now it works like `loaddata` django command with domain overwriting, right? Do we plan to make it more like a `AutorisatieForm` in the admin, so the user could choose component and `related_type_selection` for this component? Do we want to provide it? Do clients want it? Maybe it can be discussed with @joeribekker as a next iteration of this config? It seems nice to have all authorizations configured without providing a single zaaktype url. But it could be the case that no one needs it and I overthink it.
2. Please add this step to the web_init in docker-compose file. It can be then shown to devops as an example how to configure this step

**Comment by stevenbal** [comment]
@annashamray 

> For now it works like loaddata django command with domain overwriting, right? 

that's right, it mostly relies on loaddata

> Do we plan to make it more like a AutorisatieForm in the admin, so the user could choose component and related_type_selection for this component?

I think this should already be possible, because the `AutorisatieSpec`s are exported/imported as well?

> Please add this step to the web_init in docker-compose file. It can be then shown to devops as an example how to configure this step

will do :+1:

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [reply src/openzaak/config/bootstrap/authorizations.py:71]: I have discussed it with Joeri, it could indeed be a single target domain, but the thing is that we do support the usage of external catalogi as well, so if we have an import file that contains multiple different domains, the import wouldn't hold up.

I do still wonder if this is actually used anywhere in practice though, maybe we should give two options to simplify the config for instances that don't use external catalogi:
- either supply a `AUTHORIZATIONS_CONFIG_TARGET_DOMAIN` and just map all URLs to that domain (this assumes that the import file does not contain any external catalogi
- or supply a mapping file in case there are external catalogi?

File: src/openzaak/management/commands/dump_auth_fixtures.py
Comment (stevenbal) [reply src/openzaak/management/commands/dump_auth_fixtures.py:1]: Hmm yeah I'll move it to a bash script, though I do think at some point we'll want to be able to export from the admin interface, but we can just do `call_command` with `dumpdata` in that case

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (stevenbal) [reply src/openzaak/config/tests/bootstrap/test_load_authorizations.py]: I removed the tempfiles and added two static .yaml files

**Comment by annashamray** [comment]
@joeribekker @stevenbal Should we park this issue until https://github.com/open-zaak/open-zaak/issues/1661 is implemented? I see that nobody is assigned to it, what is our plan?

**Comment by stevenbal** [comment]
@annashamray yeah I think it would be better to wait for that. I think we're waiting for approval by Den Haag for the estimation

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (stevenbal) [inline src/openzaak/config/tests/bootstrap/test_load_authorizations.py:163]: I'm not sure what else we can do in case the Catalogus does not exist, not sure if we can avoid errors from being raised because we use `loaddata`

File: docs/installation/config/openzaak_config_cli.rst
Comment (stevenbal) [inline docs/installation/config/openzaak_config_cli.rst:109]: Can we rely on this existing envvar? Or should I add a new one for this?

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (stevenbal) [reply src/openzaak/config/tests/bootstrap/test_load_authorizations.py:163]: On second thought, it currently will only be possible to run this step after catalogi have been configured (be it manually or via some fixture or imports), because otherwise this step will fail

File: bin/setup_configuration.sh
Comment (stevenbal) [inline bin/setup_configuration.sh:12]: I added this because otherwise it won't be possible to run the setup configuration part for authorizations, because the catalogi do not exist yet

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [inline src/openzaak/config/bootstrap/authorizations.py:86]: @annashamray I added this to allow using the fixture as a single source of truth, but I just realized that other ConfigurationSteps also add applicaties and jwtsecrets. How should I treat those? I guess I could exclude the stuff created in previous steps from the .delete, but I wanted to check with you before I make it more complicated

**Comment by stevenbal** [comment]
I rebased this PR to work with CatalogusAutorisaties, so it is ready for review, though I do have several questions relating to the implementation in the PR

File: docs/installation/config/openzaak_config_cli.rst
Comment (annashamray) [reply docs/installation/config/openzaak_config_cli.rst:109]: Let's reuse it, I think it should be fine

File: src/openzaak/config/bootstrap/authorizations.py
Comment (annashamray) [reply src/openzaak/config/bootstrap/authorizations.py:86]: It can be a leaky abstraction. but I'd check if other steps which add autorisation data are enabled. 
If they are enabled and AUTHORIZATIONS_CONFIG_DELETE_EXISTING = True, I'd raise an Exception

It's first time then we have configuration steps that can't be used together. Maybe in the future we can add some steprelation validator that can check which will have some mapping of steps that can't be run together?

File: src/openzaak/config/bootstrap/authorizations.py
Comment (annashamray) [inline src/openzaak/config/bootstrap/authorizations.py:112]: My favorite question - should we have an additional validation for the fixture? For example, checking if zaaktypen are existing urls or checking that catalog autorisatie and individual autorisatie don't overlap? 

Maybe we can do it in the `test_configuration` part?

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (annashamray) [reply src/openzaak/config/tests/bootstrap/test_load_authorizations.py:163]: I think raising `ConfigurationRunFailed` if the auth fixture is not correct is an expected behaviour :+1:

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (annashamray) [reply src/openzaak/config/tests/bootstrap/test_load_authorizations.py:257]: I think you already solved it by introducing `AUTHORIZATIONS_CONFIG_DELETE_EXISTING` env var

**Review by annashamray — COMMENTED** [review]
Nice code and documentation!

File: src/openzaak/config/tests/bootstrap/test_load_authorizations.py
Comment (stevenbal) [reply src/openzaak/config/tests/bootstrap/test_load_authorizations.py:163]: I think so too, and with the change to `bin/setup_configuration.sh` to load fixtures first if `AUTHORIZATIONS_CONFIG_ENABLE` is True, this is no longer an issue

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [reply src/openzaak/config/bootstrap/authorizations.py:91]: Since the mapping file construction is already there and documented, I'll leave it in

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [reply src/openzaak/config/bootstrap/authorizations.py:112]: I think validating it might be a good idea, though I'm not sure if `test_configuration` is the right place, since that runs after `configure`, so you'd be left with a selftestfailure, but the configuration was loaded.

I'm not sure if it's easy to keep all the validation between the admin interface, API and this configuration step in sync, I'll see what I can do with it

Comment (stevenbal) [reply src/openzaak/config/bootstrap/authorizations.py:112]: I've been trying to do this, but I don't think it's possible to share the same validation between the admin form and the configuration step without duplicating it unfortunately, particularly validating the overlap between autorisaties/catalogusautorisaties is annoying, because the fixture has a different shape for the data than the AutorisatieFormSet (where multiple autorisaties are grouped together). To further complicate things, in the form validation `zaaktypen/iotypen/besluittypen` are resolved using django loose fk, while in the configurationstep this has to be done manually as well

@annashamray do you have any ideas? :grimacing:

File: src/openzaak/config/bootstrap/authorizations.py
Comment (stevenbal) [reply src/openzaak/config/bootstrap/authorizations.py]: I added code to raise an exception if any of OPENZAAK_NOTIF_CONFIG_ENABLE, NOTIF_OPENZAAK_CONFIG_ENABLE, DEMO_CONFIG_ENABLE are True

File: src/openzaak/config/bootstrap/authorizations.py
Comment (annashamray) [reply src/openzaak/config/bootstrap/authorizations.py:166]: Let's park this for now and see how this configuration step is used to determine the next step

**Review by annashamray — APPROVED** [review]
Good job!

**Pull request merged**
